includes/class-wc-payments-order-service.php

```diff
@@ -1240,6 +1240,67 @@ public function capture_authorization_on_order_status_change( int $order_id ) {
 		}
 	}
 
+	/**
+	 * Creates a refund for the given order.
+	 *
+	 * @param WC_Order $order The order to refund.
+	 * @param float    $amount The amount to refund.
+	 * @param string   $refund_id The refund ID.
+	 * @param string   $reason The reason for the refund.
+	 * @param array    $line_items The line items to refund.
+	 *
+	 * @throws Exception If the refund creation fails.
+	 */
+	public function create_refund_for_order( WC_Order $order, float $amount, string $refund_id, string $reason = '', array $line_items = [] ) {
+		$refund_params = [
+			'amount'    => $amount,
+			'reason'    => $reason,
+			'order_id'  => $order->get_id(),
+			'refund_id' => $refund_id,
+		];
+
+		if ( $line_items ) {
+			$refund_params['line_items'] = $line_items;
+		}
+
+		$refund = wc_create_refund(
+			$refund_params
+		);
+
+		if ( is_wp_error( $refund ) ) {
+			throw new Exception( $refund->get_error_message() );
+		}
+
+		return $refund;
+	}
+
+	/**
+	 * Processes a refund for the given order.
+	 *
+	 * @param WC_Order        $order The order to refund.
+	 * @param WC_Order_Refund $wc_refund The WC refund object.
+	 * @param string          $refund_id The refund ID.
+	 * @param string|null     $refund_balance_transaction_id The balance transaction ID of the refund.
+	 * @throws Order_Not_Found_Exception
+	 * @throws Exception
+	 */
+	public function process_order_refund( WC_Order $order, WC_Order_Refund $wc_refund, string $refund_id, ?string $refund_balance_transaction_id ): void {
+		$note = ( new WC_Payments_Refunded_Event_Note( $wc_refund->get_amount(), $wc_refund->get_currency(), $refund_id, $wc_refund->get_reason(), $order ) )->generate_html_note();
+
+		if ( ! $this->order_note_exists( $order, $note ) ) {
+			$order->add_order_note( $note );
+		}
+
+		// Set refund metadata.
+		$this->set_wcpay_refund_status_for_order( $order, 'successful' );
+		$this->set_wcpay_refund_id_for_order( $order, $refund_id );
```

Comment:
The original code used param `$wc_last_refund`: 

```suggestion
		$this->set_wcpay_refund_id_for_order( $wc_refund, $refund_id );
```

--- 

I think the method name `set_wcpay_refund_id_for_order` is a bit confusing, as this needs to apply to `WC_Order_Refund` object rather than `WC_Order` object. 

Actually, the meta key `_wcpay_refund_id` for this method was originally introduced here https://github.com/Automattic/woocommerce-payments/pull/3965. And so far, the purpose of this meta key is still the same, i.e. dealing with async failed refund event. 

It would be nice if this PR can also change the name of `set_wcpay_refund_status_for_order` to `set_wcpay_refund_id_for_refund`

Reply:
Oops I did not notice, it seems indeed confusing, thanks for catching it. Addressed the change of name here https://github.com/Automattic/woocommerce-payments/pull/8200/commits/ebdd3df9798513922fa767079651bfa1b0f57a25 thanks!